src/librbd/mirror/snapshot/CreatePrimaryRequest2.cc

```diff
@@ -0,0 +1,507 @@
+        if (info == nullptr) {
+          continue;
+        }
+        if (info->state != cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY) {
+          // reset counters -- we count primary snapshots after the last
+          // promotion
+          count = 0;
+          unlink_snap_id = 0;
+          continue;
+        }
+        if (info->mirror_peer_uuids.count(peer) == 0) {
+          // snapshot is not linked with this peer
+          continue;
+        }
+        if (info->group_spec.is_valid() || !info->group_snap_id.empty()) {
+          // snap is part of a group snap
+          continue;
+        }
+        count++;
+        if (count == max_snapshots) {
+          unlink_snap_id = snap_info_pair.first;
+        }
+        if (count > max_snapshots) {
+          peer_uuid = peer;
+          snap_id = unlink_snap_id;
+          goto do_unlink;
+        }
+      }
+    }
+  }
+
+  finish(0);
+  return;
+
+do_unlink:
+  ldout(m_cct, 15) << "peer=" << peer_uuid << ", snap_id=" << snap_id << dendl;
+
+  auto ctx = create_context_callback<
+    CreatePrimaryRequest2<I>,
+    &CreatePrimaryRequest2<I>::handle_unlink_peer>(this);
+  auto req = UnlinkPeerRequest<I>::create(m_image_ctxs[0], snap_id, peer_uuid, true,
+                                          ctx);
+  req->send();
+}
+
+template <typename I>
+void CreatePrimaryRequest2<I>::handle_unlink_peer(int r) {
+  ldout(m_cct, 15) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to unlink peer: " << cpp_strerror(r) << dendl;
+    finish(0); // not fatal
+    return;
+  }
+
+  unlink_peer();
+}
+
+
+template <typename I>
+void CreatePrimaryRequest2<I>::notify_quiesce() {
+
+  if (m_group_snap_id.empty()) {
+    create_snapshot();  
+    return;
+  }
+  if ((m_snap_create_flags & SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE) != 0) {
+    acquire_exclusive_locks();
+    return;
+  }
+  ldout(m_cct, 15) << dendl;
+
+  // The individual image snaps do not need to quiesce.
+  m_snap_create_flags |= SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE;
+  auto ctx = create_context_callback<
+    CreatePrimaryRequest2<I>,
+    &CreatePrimaryRequest2<I>::handle_notify_quiesce>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  int image_count = m_image_ctxs.size();
+  m_quiesce_requests.resize(image_count);
+
+  for (int i = 0; i < image_count; ++i) {
+    auto ictx = (m_image_ctxs)[i];
+    ictx->image_watcher->notify_quiesce(&(m_quiesce_requests)[i], m_prog_ctx,
+                                        gather_ctx->new_sub());
+  }
+
+  gather_ctx->activate();
+
+}
+
+
+template <typename I>
+void CreatePrimaryRequest2<I>::handle_notify_quiesce(int r) {
+  ldout(m_cct, 15) << "r=" << r << dendl;
+
+  if (r < 0 &&
+      (m_snap_create_flags & SNAP_CREATE_FLAG_IGNORE_NOTIFY_QUIESCE_ERROR) == 0) {
+    m_ret_code = r;
+    notify_unquiesce();
+    return;
+  }
+  acquire_exclusive_locks();
+
+}
+
+
+template <typename I>
+void CreatePrimaryRequest2<I>::notify_unquiesce() {
```

Comment:
`notify_unquiesces`?

Reply:
I think notify_unquiesce works for more than one image as well.